test/e2e/app/app.go

```diff
@@ -259,12 +261,20 @@ func (app *Application) ApplySnapshotChunk(req abci.RequestApplySnapshotChunk) a
 
 func (app *Application) PrepareProposal(
 	req abci.RequestPrepareProposal) abci.ResponsePrepareProposal {
-	return abci.ResponsePrepareProposal{BlockData: req.BlockData}
+	dataHash := types.ToTxs(req.Txs).Hash()
+	squareSize := len(req.Txs)
+	if squareSize == 0 {
+		squareSize = 1
+	}
+	squareSizeBytes := make([]byte, 8)
+	binary.BigEndian.PutUint64(squareSizeBytes, uint64(squareSize))
+	req.Txs = append(req.Txs, dataHash, squareSizeBytes)
+	return abci.ResponsePrepareProposal{Txs: req.Txs}
```

Comment:
[nit] can the placeholders here look identically to PrepareProposal in abci/types/application.go